src/Microsoft.Azure.WebJobs.Host/Loggers/Logger/ReadOnlyScopeDictionary.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Extensions.Logging
+{
+    /// <summary>
+    /// The base class for all readonly dictionaries used for scopes
+    /// </summary>
+    public class ReadOnlyScopeDictionary : IReadOnlyDictionary<string, object>
```

Comment:
Is there a way we can use a `ReadOnlyDictionary` as a backing store for this rather than implementing `IReadOnlyDictionary`? Code like this makes me nervous unless there's a whole bunch of tests validating it.

Reply:
I didn't try it. My assumption was to reduce allocation on the repeated array of keys. With scopes of 8 keys, we have an overhead for storing 8 keys + length + other fields in a dictionary that for a scope might be simply unneeded. [ReadOnlyDictionary](https://github.com/dotnet/runtime/blob/master/src/coreclr/src/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyDictionary.cs#L24) suffers from the same overhead as it wraps the dictionary.

We have benchmarks for some scenarios now. I understand that you might be unsure with this implementation. Let me give it a spin. Maybe I can replace it with a smaller alternative that isn't as untested as this one.

Reply:
A fast try out breaks on `StartTelemetryIfFunctionInvocation` which requires the state to implement `IReadOnlyDictionary` for retrieval of the data.

I see two choices: 

1. add tests
1. undo the work with the scope as the new dictionary and rerun benchmarks to see how much it impacts the performance gain. Maybe with initializing capacity properly it would just work.